File: exercise/ts/day12/tests/santa.spec.ts
import {Toy} from "../src/gifts/Toy";
import {Child} from "../src/gifts/Child";
import {Santa} from "../src/gifts/Santa";

describe("Santa's gift selection process", () => {
    const Playstation = new Toy('playstation');
    const Ball = new Toy('ball');
    const Plush = new Toy('plush');

    it('should give the third choice to a naughty child', () => {
        const bobby = new Child('bobby', 'naughty');
        bobby.setWishlist(Playstation, Plush, Ball);

        const santa = new Santa();
        santa.addChild(bobby);

        expect(santa.chooseToyForChild('bobby')).toBe(Ball);
    });

    it('should give the second choice to a nice child', () => {
        const bobby = new Child('bobby', 'nice');
        bobby.setWishlist(Playstation, Plush, Ball);

        const santa = new Santa();
        santa.addChild(bobby);

        expect(santa.chooseToyForChild('bobby')).toBe(Plush);
    });

    it('should give the first choice to a very nice child', () => {
        const bobby = new Child('bobby', 'very nice');
        bobby.setWishlist(Playstation, Plush, Ball);

        const santa = new Santa();
        santa.addChild(bobby);

        expect(santa.chooseToyForChild('bobby')).toBe(Playstation);
    });

    it('should throw an exception if the child does not exist', () => {
        const santa = new Santa();
        const bobby = new Child('bobby', 'very nice');
        bobby.setWishlist(Playstation, Plush, Ball);

        santa.addChild(bobby);

        expect(() => santa.chooseToyForChild('alice')).toThrowError('No such child found');
    });
});

File: exercise/scala/day12/src/test/scala/SantaSpec.scala
import org.scalatest.funspec.AnyFunSpec
import org.scalatest.matchers.should.Matchers
import gifts._

class SantaSpec extends AnyFunSpec with Matchers {
  val Playstation: Toy = Toy("playstation")
  val Ball: Toy = Toy("ball")
  val Plush: Toy = Toy("plush")

  describe("Santa's gift selection process") {
    it("should give the third choice to a naughty child") {
      val bobby = new Child("bobby", "naughty")
      bobby.setWishlist(Playstation, Plush, Ball)
      val santa = new Santa()
      santa.addChild(bobby)

      santa.chooseToyForChild("bobby") shouldEqual Some(Ball)
    }

    it("should give the second choice to a nice child") {
      val bobby = new Child("bobby", "nice")
      bobby.setWishlist(Playstation, Plush, Ball)
      val santa = new Santa()
      santa.addChild(bobby)

      santa.chooseToyForChild("bobby") shouldEqual Some(Plush)
    }

    it("should give the first choice to a very nice child") {
      val bobby = new Child("bobby", "very nice")
      bobby.setWishlist(Playstation, Plush, Ball)
      val santa = new Santa()
      santa.addChild(bobby)

      santa.chooseToyForChild("bobby") shouldEqual Some(Playstation)
    }

    it("should throw an exception if the child does not exist") {
      val santa = new Santa()
      val bobby = new Child("bobby", "very nice")
      bobby.setWishlist(Playstation, Plush, Ball)
      santa.addChild(bobby)

      an[NoSuchElementException] should be thrownBy santa.chooseToyForChild("alice")
    }
  }
}
